### PR TITLE
Support for two-wire cJTAG OSCAN1 signaling thru FTDI devices with appropriate pinout 

### PR DESCRIPTION
**Pull request by GregSavin**
cJTAG OSCAN1, in lieu of 4-wire JTAG, is starting to be a configuration option for some SiFive hardware.  An FTDI-based adapter that can be configured to drive the bidirectional pin TMSC is assumed for this topology.  Specifically, the Olimex ARM-USB-TINY-H with the ARM-JTAG-SWD adapter, connected to a SiFive cJTAG-enabled target board is the only known concrete topology, currently.  But in theory, other FTDI based devices that can drive a two-wire bidirectional signaling pattern could be made to work in this scheme in the future.

These code changes are offered as a way to drive that topology.  It's translating IR/DR and JTAG traversal commands to the two-wire clocking and signaling.

### CODE REVIEW

File: src/jtag/drivers/oscan1_ftdi.c
Comment (timsifive) [inline src/jtag/drivers/oscan1_ftdi.c]: I'm not sure what the copyright should be, but this doesn't look right. In the RISC-V specific code I've not bothered with a copyright line at all, nor with the GPL info. Upstream has accepted that without comment.

File: src/jtag/drivers/oscan1_ftdi.c
Comment (timsifive) [inline src/jtag/drivers/oscan1_ftdi.c]: I would remove the comment that's identical to the one in ftdi.c, and add a line to this paragraph something like:
See ftdi.c for more comments about these JTAG adapters.

**Review by timsifive — CHANGES_REQUESTED** [review]
oscan1_ftdi.c duplicates way too much of ftdi.c.
ftdi.c is 1289 lines
oscan1_ftdi.c is 1538
The diff between them inserts 415 lines and deletes 166 lines. And at least half of those simply change ftdi to oscan1 without any functional change at all.

The simplest solution is probably to add to ftdi.c what you need added there (checking for FTDI_OSCAN1, probably) and reuse as much of the ftdi.c as you currently are. The other option is to make 2 new files: one for the functions shared between ftdi and oscan1, and one for the oscan1-specific stuff. It might be worth asking openocd-devel@lists.sourceforge.net which solution is preferred before doing a bunch of work.

Both of those solutions also have the benefit that I only need to review code that is actually new, instead of reading 1000 lines of code that haven't actually changed (and are probably solid).

**Comment by GregSavin** [comment]
Thanks Tim.  Yeah I was conflicted about modifying ftdi.c to add support a for such a niche topology, thinking there might be more resistance to upstreaming later, but modifying ftdi.c would be a much smaller diff for sure.  I'll see about refactoring my changes by adding them to ftdi.c (not a whole lot of work), and then we'll see if that looks better.

**Comment by GregSavin** [comment]
My most recent commit on this branch has incorporated the previously differentiating content of oscan1_ftdi.c into ftdi.c (establishing OSCAN1 as a non-default mode that the ftdi driver can be placed in), and removed oscan1_ftdi.c.

File: src/jtag/drivers/ftdi.c
Comment (timsifive) [inline src/jtag/drivers/ftdi.c]: missing a verb?

File: src/jtag/drivers/ftdi.c
Comment (timsifive) [inline src/jtag/drivers/ftdi.c]: Can these two be const?

Comment (timsifive) [reply src/jtag/drivers/ftdi.c]: Might want to make them static, as well.

File: src/jtag/drivers/ftdi.c
Comment (timsifive) [inline src/jtag/drivers/ftdi.c]: DEBUG_JTAG_IO might be more suitable than LOG_DEBUG. Its output is only displayed with -d4, and only compiled in if the OpenOCD build is configured with --enable-verbose-jtag-io.

File: src/jtag/drivers/ftdi.c
Comment (timsifive) [inline src/jtag/drivers/ftdi.c]: Ditto.

File: src/jtag/drivers/ftdi.c
Comment (timsifive) [inline src/jtag/drivers/ftdi.c:1019]: Do you need to return here if any of these signals is not defined?

File: src/jtag/drivers/ftdi.c
Comment (timsifive) [inline src/jtag/drivers/ftdi.c]: If possible, can you emphasize that this is not some SiFive-specific invention, but there may be other hardware out there that this code is useful for?

File: src/jtag/drivers/ftdi.c
Comment (timsifive) [inline src/jtag/drivers/ftdi.c]: I love the comments here, but I think the compiled code would be much smaller if you had something like:
sequence[] = {{1,1,0},
{1,1,1},
...};

and then a loop to iterate over the sequence and call oscan1_set_tck_tms_tdi for each triple.

But I don't feel strongly, and I'm not convinced upstream would care too much either.

File: src/jtag/drivers/ftdi.c
Comment (timsifive) [inline src/jtag/drivers/ftdi.c]: There's a COMMAND_PARSE_ON_OFF macro that will do this for you.

**Review by timsifive — CHANGES_REQUESTED** [review]
Overall this looks good. I have a bunch of feedback, but this feels really close to what we want.

This code violates the OpenOCD coding standards in all kinds of ways. The Travis build is supposed to flag that, but somehow doesn't. I'll fix that. In your own checkout can you run:
```git diff -U20 HEAD~40 | ./tools/scripts/checkpatch.pl --no-signoff -```
and fix whatever it complains about?

**Comment by GregSavin** [comment]
Thanks @timsifive , the latest commits to this branch are intended to resolve your most recent set of comments (first a commit to address the non-formatting things and then a separate commit that addressed formatting problems flagged by checkpatch).

**Review by timsifive — APPROVED** [review]

**Comment by GregSavin** [comment]
Thanks @timsifive for reviewing/approving.  At this point, do I need to take further action for this to be merged?  Or is that something that somebody other than myself would do?

**Comment by timsifive** [comment]
My workflow is:
1. Create PR.
2. Get review approval.
3. Click "Squash and merge"

You're welcome to do the same.

In addition, I try to "Squash and merge" any changes that I've approved after giving the 24 hours for other people to comment on.

**Pull request merged**
